File: RCode/Chapter4/temporalAbstraction.R
##########################################################################
#
#   Mark Hoogendoorn & Burkhardt Funk (2017)
#   Machine Learning for the Quantified Self
#   Springer
#   Chapter 4 - Feature Engineering based on Sensory Data
#   ./chapter4/temporalAbstraction.R
# 
##########################################################################

# This function aggregates a list of values using the specified aggregation
# function (which can be 'mean', 'max', 'min', 'median', 'std', 'slope')
aggregateValue = function(data, aggregationFunction) {
  if (aggregationFunction == "mean") return(mean(data))
  if (aggregationFunction == "max") return(max(data))
  if (aggregationFunction == "min") return(min(data))
  if (aggregationFunction == "median") return(median(data))
  if (aggregationFunction == "std") return(sd(data))
  if (aggregationFunction == "slope") return(as.numeric(lm(data~(x=1:length(data)))$coef[2]))
  stop(paste("aggregationFunction:",aggregationFunction,"not found"))
}


abstractNumerical = function(df, cols, windowSize, aggregationFunction) {
  for (col in cols) {
    newColName = paste(col,"_temp_",aggregationFunction,"_ws_",windowSize,sep = "")
    data = df[,col]
    aggData = sapply((windowSize+1):length(data), function(x) {aggregateValue(data[(x-windowSize):x],aggregationFunction)})
    df[floor(windowSize/2)+(1:(length(data)-windowSize)),newColName] = aggData
  }
  return(df)
}

# Function to abstract categorical data. Note that we assume a list of binary columns representing
# the different categories 
abstractCategorical = function(df, cols, match, minSupport, windowSize, maxPatternSize) {
  # find all the relevant columns of binary attributes.
  colNames = names(df)
  selectedPatterns = c()
  relevantCols = c()
  
  for (i in 1:length(cols)) {
    if(match[i]=="exact") relevantCols = c(relevantCols,cols[i]) else
      relevantCols = c(relevantCols, colNames[grepl(cols[i],colNames)])
  }
  
  # generate the one patterns first
  potential1Patterns = as.list(relevantCols)
  temp = selectKPatterns(df,potential1Patterns,minSupport,windowSize)
  df = temp[[1]]
  onePatterns = temp[[2]]
  selectedPatterns[[length(selectedPatterns)+1]] = onePatterns
  print(paste("Number of patterns of size 1:",length(onePatterns)))

  k = 1
  kPatterns = onePatterns
  while(k< maxPatternSize & length(kPatterns)>0) {
    k = k + 1
    potentialKPatterns = extendKPatterns(kPatterns,onePatterns)
    print(potentialKPatterns)
    temp = selectKPatterns(df, potentialKPatterns, minSupport, windowSize)
    df = temp[[1]]
    kPatterns = temp[[2]]
    selectedPatterns[[length(selectedPatterns)+1]] = kPatterns
    print(paste("Number of patterns of size",k,":",length(kPatterns)))
  }
  return(df)
}

# selects the patterns from 'patterns' that meet the minimum support in the dataset
# given the window size.
selectKPatterns = function(df, patterns, minSupport, windowSize){
  sP = list()
  for (pattern in patterns) {
    # determine the number of occurrences of a pattern
    times = determinePatternTimes(df, pattern, windowSize)
    support = length(times)/nrow(df)
    
    # If we meet the minum support, append the selected patterns and set the
    # value to 1 at which it occurs.
    if (support>=minSupport) {
      sP[[length(sP)+1]] = pattern
      df[,paste(c("temp_pattern",pattern),collapse = "_")] = 0
      df[times,paste(c("temp_pattern",pattern),collapse = "_")] = 1
      print(pattern)
    }
  }
  return(list(df,sP))
}

determinePatternTimes = function(df, pattern, windowSize){
  times = c()
  if (length(pattern) == 1) {
    times = which(df[pattern]>0)
  } else if(length(pattern) == 3) {
    # If we have a complex pattern (<n> (b) <m> or <n> (c) <m>)
    # due to naming conventions, brackets should not be used
    # we therefore use ".b." instead of "(b)"
    timesFirstPart = determinePatternTimes(df, pattern[1], windowSize)
    timesSecondPart = determinePatternTimes(df, pattern[3], windowSize)
    if (pattern[2] == ".c.") {
      if (pattern[1]==pattern[3]) {
        # No use for co-occurences of the same patterns
        times = c()
      } else {
        times = intersect(timesFirstPart,timesSecondPart)
      } 
    } else if (pattern[2] == ".b.") {
      times = unlist(sapply(timesFirstPart, function(t){
        if(sum(t<timesSecondPart & timesSecondPart<=t+windowSize )>0) return(t)
      }))
      
    } else stop("complex operator not known")
    
  } else stop("pattern length error")
  
  return(times)
}

# extends a set of k-patterns with the 1-patterns that have sufficient support.
extendKPatterns = function(kPatterns, onePatterns) {
  newPatterns = list()
  for(kP in kPatterns) {
    for(oneP in onePatterns) {
      newPatterns[[length(newPatterns)+1]] = c(kP,".b.",oneP)
      newPatterns[[length(newPatterns)+1]] = c(kP,".c.",oneP)
    }
  }
  return(newPatterns)
}

File: Python2Code/Chapter4/TemporalAbstraction.py
##############################################################
#                                                            #
#    Mark Hoogendoorn and Burkhardt Funk (2017)              #
#    Machine Learning for the Quantified Self                #
#    Springer                                                #
#    Chapter 4                                               #
#                                                            #
##############################################################

import numpy as np
import scipy.stats as stats

# Class to abstract a history of numerical values we can use as an attribute.
class NumericalAbstraction:


    # This function aggregates a list of values using the specified aggregation
    # function (which can be 'mean', 'max', 'min', 'median', 'std', 'slope')
    def aggregate_value(self, data, aggregation_function):
        # Compute the values and return the result.
        if aggregation_function == 'mean':
            return data.mean(skipna = True)
        elif aggregation_function == 'max':
            return data.max(skipna = True)
        elif aggregation_function == 'min':
            return data.min(skipna = True)
        elif aggregation_function == 'median':
            return data.median(skipna = True)
        elif aggregation_function == 'std':
            return data.std(skipna = True)
        elif aggregation_function == 'slope':
            # For the slope we need a bit more work.
            # We create time points, assuming discrete time steps with fixed delta t:
            times = np.array(range(0, len(data.index)))
            data = data.as_matrix().astype(np.float32)

            # Check for NaN's
            mask = ~np.isnan(data)

            # If we have no data but NaN we return NaN.
            if (len(data[mask]) == 0):
                return np.nan
            # Otherwise we return the slope.
            else:
                slope, intercept, r_value, p_value, std_err = stats.linregress(times[mask], data[mask])
                return slope
        else:
            return np.nan

    # Abstract numerical columns specified given a window size (i.e. the number of time points from
    # the past considered) and an aggregation function.
    def abstract_numerical(self, data_table, cols, window_size, aggregation_function):

        # Create new columns for the temporal data.
        for col in cols:
            data_table[col + '_temp_' + aggregation_function + '_ws_' + str(window_size)] = np.nan

        # Pass over the dataset (we cannot compute it when we do not have enough history)
        # and compute the values.
        for i in range(window_size, len(data_table.index)):
            for col in cols:
                data_table.ix[i, col + '_temp_' + aggregation_function + '_ws_' +str(window_size)] = self.aggregate_value(data_table[col][i-window_size:min(i+1, len(data_table.index))], aggregation_function)

        return data_table

# Class to perform categorical abstraction. We obtain patterns of categorical attributes that occur frequently
# over time.
class CategoricalAbstraction:

    pattern_prefix = 'temp_pattern_'
    before = '(b)'
    co_occurs = '(c)'
    cache = {}

    # Determine the time points a pattern occurs in the dataset given a windows size.
    def determine_pattern_times(self, data_table, pattern, window_size):
        times = []

        # If we have a pattern of length one
        if len(pattern) == 1:
            # If it is in the cache, we get the times from the cache.
            if self.cache.has_key(self.to_string(pattern)):
                times = self.cache[self.to_string(pattern)]
            # Otherwise we identify the time points at which we observe the value.
            else:
                timestamp_rows = data_table[data_table[pattern[0]] > 0].index.values.tolist()
                times = [data_table.index.get_loc(i) for i in timestamp_rows]
                self.cache[self.to_string(pattern)] = times

        # If we have a complex pattern (<n> (b) <m> or <n> (c) <m>)
        elif len(pattern) == 3:
            # We computer the time points of <n> and <m>
            time_points_first_part = self.determine_pattern_times(data_table, pattern[0], window_size)
            time_points_second_part = self.determine_pattern_times(data_table, pattern[2], window_size)

            # If it co-occurs we take the intersection.
            if pattern[1] == self.co_occurs:
                # No use for co-occurences of the same patterns...
                if pattern[0] == pattern[2]:
                    times = []
                else:
                    times = list(set(time_points_first_part) & set(time_points_second_part))
            # Otherwise we take all time points from <m> at which we observed <n> within the given
            # window size.
            elif pattern[1] == self.before:
                for t in time_points_second_part:
                    if len([i for i in time_points_first_part if ((i >= t - window_size) & (i < t))]):
                        times.append(t)
        return times

    # Create a string representation of a pattern.
    def to_string(self, pattern):
        # If we just have one component, return the string.
        if len(pattern) == 1:
            return str(pattern[0])
        # Otherwise, return the merger of the strings of all
        # components.
        else:
            name = ''
            for p in pattern:
                name = name + self.to_string(p)
            return name

    # Selects the patterns from 'patterns' that meet the minimum support in the dataset
    # given the window size.
    def select_k_patterns(self, data_table, patterns, min_support, window_size):
        selected_patterns = []
        for pattern in patterns:
            # Determine the times at which the pattern occurs.
            times = self.determine_pattern_times(data_table, pattern, window_size)
            # Compute the support
            support = float(len(times))/len(data_table.index)
            # If we meet the minum support, append the selected patterns and set the
            # value to 1 at which it occurs.
            if support >= min_support:
                selected_patterns.append(pattern)
                print self.to_string(pattern)
                # Set the occurrence of the pattern in the row to 0.
                data_table[self.pattern_prefix + self.to_string(pattern)] = 0
                data_table.ix[times, self.pattern_prefix + self.to_string(pattern)] = 1
        return data_table, selected_patterns


    # extends a set of k-patterns with the 1-patterns that have sufficient support.
    def extend_k_patterns(self, k_patterns, one_patterns):
        new_patterns = []
        for k_p in k_patterns:
            for one_p in one_patterns:
                # Add a before relationship
                new_patterns.append([k_p, self.before, one_p])
                # Add a co-occurs relationship.
                new_patterns.append([k_p, self.co_occurs, one_p])
        return new_patterns


    # Function to abstract our categorical data. Note that we assume a list of binary columns representing
    # the different categories. We set whether the column names should match exactly 'exact' or should include the
    # specified name 'like'. We also express a minimum support,a windows size between succeeding patterns and a
    # maximum size for the number of patterns.
    def abstract_categorical(self, data_table, cols, match, min_support, window_size, max_pattern_size):

        # Find all the relevant columns of binary attributes.
        col_names = list(data_table.columns)
        selected_patterns = []

        relevant_dataset_cols = []
        for i in range(0, len(cols)):
            if match[i] == 'exact':
                relevant_dataset_cols.append(cols[i])
            else:
                relevant_dataset_cols.extend([name for name in col_names if cols[i] in name])

        # Generate the one patterns first
        potential_1_patterns = [[pattern] for pattern in relevant_dataset_cols]

        new_data_table, one_patterns = self.select_k_patterns(data_table, potential_1_patterns, min_support, window_size)
        selected_patterns.extend(one_patterns)
        print 'Number of patterns of size 1 is ' + str(len(one_patterns))

        k = 1
        k_patterns = one_patterns

        # And generate all following patterns.
        while (k < max_pattern_size) & (len(k_patterns) > 0):
            k = k + 1
            potential_k_patterns = self.extend_k_patterns(k_patterns, one_patterns)
            new_data_table, selected_new_k_patterns = self.select_k_patterns(new_data_table, potential_k_patterns, min_support, window_size)
            selected_patterns.extend(selected_new_k_patterns)
            print 'Number of patterns of size ' + str(k) + ' is ' + str(len(selected_new_k_patterns))

        return new_data_table



